docs/rfc/rfc-8.adoc

```diff
@@ -0,0 +1,410 @@
+6-of-9 Threshold Council Multisig. Since the Gnosis Safe for Threshold Council
+was deployed using Safe Proxy factory v1.3.0, it
+link:https://help.safe.global/en/articles/5267779-i-sent-assets-to-a-safe-address-on-the-wrong-network-any-chance-to-recover[should be possible]
+to replay the same transaction creating Gnosis Safe with the same address on the
+supported L2s and sidechains.
+
+The Governance will be able to add and remove minters to L2/sidechain TBTC
+canonical contract.
+
+=== Code organization
+
+The code that exists in the `solidity` directory should contain components
+specific to L1 and generic L2 components that will be reused between EVM
+L2 and sidechain implementations. The code specific to individual chains should
+be placed in a chain-specific directory, in a separate NPM project:
+`cross-chain/{$chainName}`.
+
+For example:
+
+* `cross-chain/arbitrum`,
+* `cross-chain/polygon`,
+* `cross-chain/optimism`.
+
+Each `cross-chain` project should contain L1 and L2
+contracts specific to that individual subchain. This separation will allow us to
+abstract out the complexity of deployment and avoid redeploying all L1 testnet
+contracts in case a single change in one of L2 contracts is needed.
+
+This organization of the code will also allow us to not add subchain-specific
+dependencies to the L1 project and to deploy NPM packages separately:
+
+* `@keep-network/tbtc-v2-arbitrum`,
+* `@keep-network/tbtc-v2-polygon`,
+* `@keep-network/tbtc-v2-optimism`.
+
+Every chain requires its own network and compiler configuration.
+The `@keep-network/tbtc-v2` package is quite heavy and there is no
+straightforward way to distinguish on which chain the given contract was
+deployed if we do not separate NPM packages.
+
+Each project should have its own CI process that may incorporate jobs specific
+to that chain if needed. 
+
+The CI processes of `cross-chain` projects should include Goerli deployment
+jobs. Note that the separation of the code does not mean the deployment is fully
+separated between chains. Both L1 and L2 contracts need to be deployed from the
+given `cross-chain` module. L1 contracts may require addresses of contracts from
+L2 and L2 contract addresses may require addresses of contracts from L1.
+
+```
+├── solidity
+│   ├── (...)
+│   └── l2
+│       └── L2TBTC.sol
+└── cross-chain
+    ├── arbitrum
+    │   ├── package.json
+    │   └── solidity
+    │      ├── L1ArbitrumGateway.sol 
+    │      ├── L2ArbitrumGateway.sol 
+    │      └── ArbitrumTBTC.sol 
+    ├── optimism
+    │   ├── package.json
+    │   └── solidity
+    │      ├── L1OptimismGateway.sol 
+    │      ├── L2OptimismGateway.sol 
+    │      └── OptimismTBTC.sol 
+    └── polygon
+        ├── package.json
+        └── solidity
+           ├── L1PolygonGateway.sol 
+           ├── L2PolygonGateway.sol 
+           └── PolygonTBTC.sol 
+```
+
+`L2TBTC.sol` is an abstract contract doing all the heavy lifting: upgradeability,
```

Comment:
What do you think about landing this with a consistent contract address across EVMs? I'd love to see it happen with `CREATE2`. Also, hell... grinding some 0's on the address would save gas :grin:

Reply:
I was on the fence about it. We can't have the same token address on L1 and L2 given the code is different. We could have the same address on all L2s but I am afraid this could be more confusing to users than different addresses everywhere. We don't want someone who sees 3 identical addresses on L2 to assume the address is the same on L1. What do you think about it? I am open to `CREATE2`, the potential confusion is my one and only concern.

Reply:
I would lean towards having it at the same address everywhere. It saves you from a lot of configuration wrangling for applications. E.g. at nomad each contract address must be [recorded separately](https://github.com/nomad-xyz/config/blob/main/production.json#L474-L553) on each network

Users should (generally) not interact with the address directly unless they're relatively sophisticated, so address choice should be developer-facing and simplify devex

Reply:
I am looking at the nomad production JSON config and my understanding is that in our case you would have to provide a separate address for Ethereum L1 anyway. Wouldn't it be more confusing for dev? 🤔

Reply:
@pdyraga I am also in favor of having the same address everywhere since most of the users will not check/remember the token contract address unless they are sophisticated. But I am pretty much sure that they will not remember the token contract address. I believe your concern can become reality only if they get to see the token contract addresses side by side and I do not know when this could happen, I am sure you can provide me with some examples, but to be honest I don't think it would be so bad. 

To avoid the DEV confusion just signal the chain name in front of the addresses in the documentation or wherever we will have all the addresses listed together. 

I wonder if having the same token address is also helpful in the wallet token import. Just a thought, nothing solid.

Reply:
Matching contract addresses on different chains and networks seems convenient but brings a risk of confusion. Unique addresses are another level of protection for users/devs to ensure they are interacting with the right chain/network. It protects against accidentally locking funds in a contract deployed on an undesired network.

Reply:
I think this is a small enough implementation detail that we don't need to block merging this RFC on resolving it